### ISSUE TITLE
Restarting with hotkeys does not result in `vscode.onDidTerminateDebugSession` being called

### ISSUE THREAD

**Issue by bwateratmsft** [microsoft/vscode#117984]
- VS Code Version: 1.53.2
- OS Version: Win10 20H2

Steps to Reproduce:

1. In a VSCode extension, register a listener for `vscode.debug.onDidTerminateDebugSession`
1. Launch some debug session
1. Stop the session. Observe, the listener gets called.
1. Launch the session again
1. _Restart_ the session, using the button in the UI (_not_ <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F5</kbd>). Observe, the listener gets called.
1. _Restart_ the session, using the keyboard, <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F5</kbd>. Observe, the listener _does not_ get called.

Found via https://github.com/microsoft/vscode-docker/issues/2676. The Docker extension uses this handler to shut down containers left by debugging, after the debugging has stopped. Since this isn't being called, the container is left running, and debugging attempts to restart--but fails, because the container name is still in use.

**Comment by bwateratmsft** [comment]
It may be the same issue or related, but I found while investigating https://github.com/microsoft/vscode-docker/issues/2706 that `vscode.debug.onDidTerminateDebugSession` also does not get called after terminating the second session after restart.

1. Launch debug
1. Restart with the UI button (`vscode.debug.onDidTerminateDebugSession` gets called)
1. Stop with the UI button (`vscode.debug.onDidTerminateDebugSession` _does not_ get called)

I'm adjusting the title and adding this to the original comment.

**Comment by isidorn** [comment]
I think I fixed this today via https://github.com/microsoft/vscode/issues/117975 I would kindly ask you to try this again with VS Code insiders on Friday and let me know if you can still reproduce this. Thanks a lot
https://code.visualstudio.com/insiders/

**Comment by bwateratmsft** [comment]
Will do, thanks @isidorn!

**Comment by bwateratmsft** [comment]
@isidorn I just tried it out and it seems that the same problem is happening.
![image](https://user-images.githubusercontent.com/36966225/110149945-a5890780-7dac-11eb-99fe-a395cdd62557.png)

**Comment by isidorn** [comment]
@bwateratmsft thanks I just tried reproducing this using your exact steps and `js-debug` and I could not reproduce.
What I suggest you do is F1 > developer tools > put a breakpoint [here](https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/browser/debugCommands.ts#L221)

And check what is the difference when that command is being called by click (when it works) and by keybaord shortcut when it does not work. For me they are absolutely the same.

**Comment by bwateratmsft** [comment]
We don't actually have our own debug adapter but instead resolve into others; could that be related? e.g. our `docker` `launch` debug configuration for .NET resolves into a `coreclr` config; for Python into `python`, Node into `node2`, etc.

I tried out those steps and the breakpoint you specified gets hit whether I click the restart button the first time (everything works), the second time (no notification of terminated session), or use the keybinding (no notification).

**Comment by isidorn** [comment]
@bwateratmsft check when the breakpoint gets hit does the command get triggered with the same arguments - does it look like it is doing the same thing. If yes then I am pretty clueless what could cause this.

Maybe your complex setup has to do. But stlll I do not see why clicking the button vs pressing keybinding would have a difference...

**Comment by bwateratmsft** [comment]
I think I just figured out about 50% of what's going on, I'm sorry I didn't realize this earlier.

We register a listener for the `onDidTerminateDebugSession` within the call to `resolveDebugConfiguration`, which only gets called the first time--not on subsequent restarts. After the first time the session terminates, we shut down the container, and then dispose of that listener. I don't know how I missed this the first time through but the keybind _does_ behave the same as the button, in that on the first time it works, the second time it doesn't.

So the root cause of the issue is that `resolveDebugConfiguration` only gets called once--at least, only once for the Docker extension--and we dispose of that listener after the first time. I'm assuming that only-called-once is by-design, so the most appropriate fix would be for us to not necessarily dispose of that listener, or find a different way to shut down the container after debug.

Is there any info that would indicate that the request was to restart the session--and therefore, we should keep the listener--rather than to terminate the session--and therefore, we should dispose? Or is there a better way to ensure we shut down the container after debugging, while also making sure not to leave open listeners that are no longer needed?

**Comment by isidorn** [comment]
Glad you could figure it out.
Yes, VS Code sends a `restart: true` in the `disconnect` request if it is part of a restart.
Pointer in the debug adapter protocol https://github.com/microsoft/vscode/blob/main/src/vs/workbench/contrib/debug/common/debugProtocol.d.ts#L575

**Comment by isidorn** [comment]
I believe you can use a [`DebugAdapterTracker`](https://github.com/microsoft/vscode/blob/main/src/vs/vscode.d.ts#L11557) and listen on all `onWillRecieveMessage`. That way you can intercept all communication.

Maybe @weinand has a better idea on how an extension can listen if the `disconnect` is a `restart`

**Comment by bwateratmsft** [comment]
Good to know. I'll also try to remember why we create one listener per `resolveDebugConfiguration`; maybe we can just use one global listener...

**Comment by weinand** [comment]
@bwateratmsft yes, I agree with @isidorn: an `DebugAdapterTracker` is the best way to track all the relevant DAP requests across all debug extensions.
